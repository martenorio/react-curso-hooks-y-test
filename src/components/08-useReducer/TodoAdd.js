import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [{ description },handleInputChange, reset] = useForm({
        description:""
    })

    const handleSubmit = (e) => {
        e.preventDefault();
        if(description.trim().length <= 1) return;
        const newTodo = {
            id:new Date().getTime(),
            desc:description,
            done:false
        };
        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar Todo</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input className='form-control'
                    onChange={handleInputChange}
                    value={description}
                    type="text" name='description'
                    placeholder='Aprender...'
                    autoComplete='off' />
                <div className="d-grid gap-2">
                    <button className="btn btn-outline-primary mt-2 btn-block">Agregar</button>
                </div>
            </form>
        </>
    )
}
