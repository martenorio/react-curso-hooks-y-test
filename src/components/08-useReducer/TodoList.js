import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos,handleDelete,handleToggle}) => {
  return (
    <ul className='list-group list-group-flush'>
        { todos.map(
            (t,i) => <TodoListItem key={ i } todo={ t } index={ i } handleDelete={ handleDelete } handleToggle={ handleToggle } />
            )
        }
    </ul>
  )
}
