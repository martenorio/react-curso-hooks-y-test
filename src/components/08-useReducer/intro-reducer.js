const initialState = [{
    id:1,
    todo:'Compras de supermercado',
    done: false
}];

const todoReducer = (state = initialState, action) => {
    if(action?.type === "agregar"){
         return [...state, action.payload]
    }
    return state;
}

let todos = todoReducer();

// No utilizar el push en react por que muta el valor de los estado
// y puede causar un error o no hacer nada.(se pudiera utilizar en useRef)

const newTodo = {
    id:2,
    todo:'Pagar recibos de luz',
    done: false
}

const agregarTodoAction = {
    type: 'agregar',
    payload:newTodo
}

todos = todoReducer(todos, agregarTodoAction);

console.log(todos);