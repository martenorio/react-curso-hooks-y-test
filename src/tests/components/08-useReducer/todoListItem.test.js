import { shallow } from "enzyme"
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem"
import { demoTodos } from '../../fixtures/demoTodos';


describe('todoListItem tests', () => {
    const todo = demoTodos[0];
    const handleToggle = jest.fn((id) => { }) // or just jest.fn()
    const handleDelete = jest.fn((id) => { }) // or just jest.fn()
    const wrapper = shallow(
        <TodoListItem
            todo={todo}
            index={0}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
        />);

    test('should show correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })
    test('should execute handleToggle', () => {
        wrapper.find('p').simulate('click')
        wrapper.find('p').simulate('click')
        expect(handleToggle.mock.calls.length).toBe(2)
        expect(handleToggle).toHaveBeenCalledWith(todo.id)
    })
    test('should execute handleDelete', () => {
        wrapper.find('button').simulate('click')
        expect(handleDelete.mock.calls.length).toBe(1);
        expect(handleDelete).toHaveBeenCalledWith(todo.id)
    })
    test('should show the text of todo', () => {
        const textRendered = wrapper.find('p').text();
        expect(textRendered).toBe(`${1} ${todo.desc}`)
    })
    test('should have complete class', () => {
        todo.done = true;
        const wrapper = shallow(
            <TodoListItem
                todo={todo}
            />)
        console.log(wrapper.html());
        expect(wrapper.find('p').hasClass('complete')).toBe(true)
        
    })
})