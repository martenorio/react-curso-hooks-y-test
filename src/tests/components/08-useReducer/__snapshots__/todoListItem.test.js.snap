// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todoListItem tests should show correctly 1`] = `
<li
  className="list-group-item"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
     
    Aprender react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;

exports[`todoListItem tests should show successfully 1`] = `
<li
  className="list-group-item"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
     
    Aprender react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
